Add docs on recommended migration patterns (#31441)
dimagi/commcare-hq: docs/migrations_in_practice.rst

**Master should always be deployable** - If you have a PR with a migration that
requires manual handling, don’t merge it until you are prepared to handle it.

**Remember third parties** - We’ll often manage migrations manually for at least
prod and india, but third parties run environments that we can’t manage
directly. Be sure that whatever changes are necessary will be applied
automatically on these environments, though it will likely require running data
migrations during deploy. If the change may be disruptive or *requires* manual
handling, `we’ll need to communicate it out in advance <announce_>`_.

.. _announce: https://confluence.dimagi.com/display/saas/Announcing+changes+affecting+third+parties

Practical Considerations
------------------------

**Handling new data** - There’s likely a code change that writes to the database
in the new way going forward. It must be deployed after any requisite schema
changes have been implemented.

**Migrating old data** - This will be handled via a django migration, a
management command, or both. Typically, small/simple migrations are handled by a
django migration and large/complex ones use a django migration that runs a

> management command. It must also be run after any schema changes.

Reviewer: ```suggestion
management command. It cannot run until any schema changes are deployed.
```

run and go-live will not be migrated. To pick up those changes, you can run the
management command a third time after the deploy, which will ensure all user
logs have been migrated.

This is still not ideal, since for the period between go-live and this third
run, there will be missing data in the DB and that data will be in-use in the
UI. Remember also that third party environments will have the management command
run only once, on the second deploy (unless we announce this as a required
maintenance operation), which would mean their data would have a gap in it.

While this single-deploy option is tempting compared to waiting weeks to get out
a multi-deploy migration, it’s really only suitable for specific situations like
custom work and unreleased features, where we can be confident the drawbacks are
insignificant.

Best practices for data migrations in Python
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Don’t fetch all data at once** - Instead, use an iterator that streams data in
chunks (note that django queryset’s ``.iter()`` method does not do this). Some
models have their own performant getters, for others, consider

> ``queryset_to_iterator`` for SQL models, ``iter_update`` or ``IterDB`` for couch

Reviewer 2: It would be nice to also mention [`DocsIterator`](https://github.com/dimagi/commcare-hq/blob/96506d5a16921469277f7ff02ff7852aade5de00/corehq/util/doc_processor/couch.py#L11-L16) here, which is a fairly new addition. I think it's one of the most robust ways to iterate over all docs of a given doc type.

Caveat: the iteration must be done fully in a single process and cannot be resumed. Use something like `resumable_view_iterator` if you need that power.

Author: Oh nice, I wasn't aware of those utils.  Feel free to add those in yourself if you've got the energy, otherwise I'll do so on my next pass when I get to it.

**Don’t write from elasticsearch** - It’s sometimes necessary to use ES to find
the data that needs to be modified, but you should only return the ids of the
objects you need, then pull the full objects from their primary database before
modifying and writing.

**Check your assumptions** - Consider what could go wrong and encode your
assumptions about the state of the world. Eg: if you expect a field to be blank,
check that it is before overwriting. Consider what would happen if your
migration were killed in the middle - would that leave data in a bad state?
Would the migration need to redo work when run again? Couch data in particular,
since it's less structured than SQL, can contain surprising data, especially in
old documents.

**Only run migrations when needed** - All historical migrations are run whenever
a new environment is set up. This means your migrations will be run in every
future test run and in every future new production or development environment.
If your migration is only relevant to environments that already have data in the
old format, decorate it with ``@skip_on_fresh_install`` so that it is a noop for
new environments.

> **Use ``run_once_off_migration``** - This encapsulates the common pattern of a

Reviewer: ```suggestion
**Use run_once_off_migration** - This encapsulates the common pattern of a
```
Seems like bold and code formatting don't work together.
